File: src/brute_force.py
"""Brute force runner with checkpointing and multiprocessing."""

from concurrent.futures import ProcessPoolExecutor, as_completed
from threading import Lock
import time

from .config import Config
from .generator import PolynomialGenerator
from .dependency_finder import DependencyFinder
from .divisibility import DivisibilityChecker
from .cache import ResultCache
from .state import BruteForceState


def process_pair_worker(args):
    """
    Worker function for processing a single pair (must be top-level for pickling).
    
    Args:
        args: Tuple of (f, g, config)
        
    Returns:
        Tuple of (f, g, q, divisibility, was_trivial, was_cached)
    """
    f, g, config = args
    
    # Create process-local instances
    cache = ResultCache(config.cache_file)
    
    try:
        # Check cache first
        cached = cache.get_result(f, g)
        if cached:
            return (f, g, None, {}, False, True)
        
        # Create instances for this process
        finder = DependencyFinder(config)
        checker = DivisibilityChecker()
        
        # Find dependency (returns tuple: q, was_trivial)
        q, was_trivial = finder.find_dependency(f, g)
        
        # Check divisibility if dependency found
        divisibility = {}
        if q:
            divisibility = checker.check_conditions(q, f, g)
        
        return (f, g, q, divisibility, was_trivial, False)
    finally:
        cache.close()


class BruteForceRunner:
    """Run brute force search with state persistence and multiprocessing."""
    
    def __init__(self, config: Config):
        """
        Initialize brute force runner.
        
        Args:
            config: Configuration object
        """
        self.config = config
        self.generator = PolynomialGenerator(config)
        self.state = BruteForceState.load(config.state_file)
        
        # Thread-safe lock for state updates
        self.state_lock = Lock()
        self.cache_lock = Lock()
        self.checkpoint_counter = 0
    
    def run(self):
        """Run brute force search with parallel processing and checkpointing."""
        print("Starting brute force search...")
        print(f"Configuration:")
        print(f"  Max degree f: {self.config.max_degree_f}")
        print(f"  Max degree g: {self.config.max_degree_g}")
        print(f"  Max degree q: {self.config.max_degree_q}")
        print(f"  Coefficient range: [{self.config.coeff_min}, {self.config.coeff_max}]")
        print(f"  Strategy: {self.config.enum_strategy}")
        print(f"  Workers: {self.config.num_workers}")
        print()
        
        if self.state.total_pairs_checked > 0:
            print("Resuming from previous state:")
            print(self.state.get_summary())
            print()
        
        try:
            # Collect pairs into batches for parallel processing
            pairs_batch = []
            
            for f, g in self.generator.generate_pairs():
                pairs_batch.append((f, g))
                
                # Process batch when it reaches batch_size
                if len(pairs_batch) >= self.config.batch_size:
                    self._process_batch(pairs_batch)
                    pairs_batch = []
            
            # Process remaining pairs
            if pairs_batch:
                self._process_batch(pairs_batch)
        
        except KeyboardInterrupt:
            print("\n\nInterrupted by user. Saving state...")
            self.state.save(self.config.state_file)
            print("State saved. You can resume later with --resume flag.")
        
        finally:
            # Final save
            self.state.save(self.config.state_file)
            
            # Print final statistics
            print("\n" + "="*60)
            print("FINAL STATISTICS")
            print("="*60)
            print(self.state.get_summary())
            
            # Get statistics with temporary cache connection
            cache = ResultCache(self.config.cache_file)
            try:
                stats = cache.get_statistics()
                print(f"\nDetailed Statistics:")
                print(f"  Total pairs checked: {stats['total']}")
                print(f"  Dependencies found: {stats['with_dependency']}")
                print(f"    - Trivial (rejected): {stats['trivial_rejected']}")
                print(f"    - Non-trivial (kept): {stats['nontrivial_found']}")
                print(f"  No dependency found: {stats['no_dependency']}")
                print(f"\nDivisibility Results (for non-trivial dependencies):")
                print(f"  ∂q/∂f : ∂q/∂x only: {stats['df_divisible_only']}")
                print(f"  ∂q/∂g : ∂q/∂x only: {stats['dg_divisible_only']}")
                print(f"  Both conditions satisfied: {stats['both_divisible']}")
            finally:
                cache.close()
    
    def _process_batch(self, pairs_batch):
        """
        Process a batch of pairs in parallel using multiple processes.
        
        Args:
            pairs_batch: List of (f, g) tuples
        """
        # Prepare arguments for worker processes
        worker_args = [(f, g, self.config) for f, g in pairs_batch]
        
        # Use ProcessPoolExecutor for true parallelism
        with ProcessPoolExecutor(max_workers=self.config.num_workers) as executor:
            # Submit all pairs for processing
            future_to_pair = {
                executor.submit(process_pair_worker, args): args[:2]
                for args in worker_args
            }
            
            # Process results as they complete
            for future in as_completed(future_to_pair):
                f, g = future_to_pair[future]
                
                try:
                    f_result, g_result, q, divisibility, was_trivial, was_cached = future.result()
                    
                    if was_cached:
                        print(f"[CACHED] f={f}, g={g}")
                    else:
                        print(f"[CHECKED] f={f}, g={g}")
                        
                        if was_trivial:
                            print(f"  Found dependency but REJECTED as trivial (only linear x)")
                        elif q:
                            print(f"  Found q={q}")
                            print(f"  ∂q/∂f : ∂q/∂x = {divisibility['df_divisible']}")
                            print(f"  ∂q/∂g : ∂q/∂x = {divisibility['dg_divisible']}")
                        else:
                            print(f"  No dependency found")
                        
                        # Save result with thread-local cache connection
                        with self.cache_lock:
                            cache = ResultCache(self.config.cache_file)
                            try:
                                cache.save_result(f, g, q, divisibility, was_trivial)
                            finally:
                                cache.close()
                        
                        # Update state (thread-safe)
                        with self.state_lock:
                            self.state.update_progress(0, 0, q is not None)
                            self.checkpoint_counter += 1
                            
                            # Checkpoint
                            if self.checkpoint_counter >= self.config.checkpoint_interval:
                                self.state.save(self.config.state_file)
                                self.checkpoint_counter = 0
                                print(f"[CHECKPOINT] {self.state.total_pairs_checked} pairs checked")
                                print()
                
                except Exception as e:
                    print(f"[ERROR] Failed to process f={f}, g={g}: {e}")

File: src/brute_force.cpp
#include "brute_force.h"
#include "cache.h"
#include <iostream>
#include <vector>
#include <sstream>
#include <unistd.h>
#include <csignal>

BruteForceRunner* BruteForceRunner::instance_ = nullptr;

BruteForceRunner::BruteForceRunner(const Config& config)
    : config_(config),
      state_(BruteForceState::load(config.state_file)),
      shutdown_requested_(false),
      generator_(nullptr),
      worker_pool_(nullptr),
      checkpoint_manager_(nullptr) {
    
    if (config_.num_workers == 0) {
        config_.num_workers = sysconf(_SC_NPROCESSORS_ONLN);
    }
    config_.batch_size = 1000 * config_.num_workers;
    
    generator_ = std::make_unique<PolynomialGenerator>(config_, shutdown_requested_, state_.f_coeffs_state, state_.g_coeffs_state);
    worker_pool_ = std::make_unique<WorkerPool>(config_, shutdown_requested_);
    checkpoint_manager_ = std::make_unique<CheckpointManager>(state_, config_.state_file, config_.checkpoint_interval);
    
    instance_ = this;
    setup_signal_handlers();
}

void BruteForceRunner::setup_signal_handlers() {
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);
}

void BruteForceRunner::signal_handler(int signal) {
    if (signal == SIGINT || signal == SIGTERM) {
        if (instance_) {
            instance_->shutdown_requested_.store(true);
        }
    }
}

void BruteForceRunner::run() {
    std::cout << "Starting brute force search..." << std::endl;
    std::cout << "Configuration:" << std::endl;
    std::cout << "  Max degree f: " << config_.max_degree_f << std::endl;
    std::cout << "  Max degree g: " << config_.max_degree_g << std::endl;
    std::cout << "  Max degree q: " << config_.max_degree_q << std::endl;
    std::cout << "  Coefficient range: [" << config_.coeff_min << ", " << config_.coeff_max << "]" << std::endl;
    std::cout << "  Workers: " << config_.num_workers << std::endl;
    std::cout << "  Batch size: " << config_.batch_size << std::endl;
    
    if (state_.total_pairs_checked > 0) {
        std::cout << "Resuming from previous state:" << std::endl;
        std::cout << state_.get_summary() << std::endl;
    }
    
    std::cout << "Starting to process polynomial pairs..." << std::endl;
    std::vector<std::pair<std::string, std::string>> batch;
    bool shutdown_msg_printed = false;
    
    generator_->generate_pairs([&](const GiNaC::ex& f, const GiNaC::ex& g) {
        if (shutdown_requested_.load()) {
            if (!shutdown_msg_printed) {
                shutdown_msg_printed = true;
            }
            return;
        }
        
        std::ostringstream f_oss, g_oss;
        f_oss << f;
        g_oss << g;
        batch.emplace_back(f_oss.str(), g_oss.str());
        
        if (batch.size() >= static_cast<size_t>(config_.batch_size)) {
            if (shutdown_requested_.load()) {
                if (!shutdown_msg_printed) {
                    shutdown_msg_printed = true;
                }
                return;
            }
            
            // Save state BEFORE processing to avoid losing data on crash
            auto batch_size = batch.size();
            state_.total_pairs_generated += batch_size;
            state_.total_pairs_checked += batch_size;
            state_.f_coeffs_state = generator_->get_f_coeffs();
            state_.g_coeffs_state = generator_->get_g_coeffs();
            state_.last_f_index = generator_->get_f_index();
            state_.last_g_index = generator_->get_g_index();
            
            checkpoint_manager_->increment_and_save_if_needed();
            
            // Now process the batch
            worker_pool_->process_batch(batch);
            batch.clear();

            if (shutdown_requested_.load()) {
                if (!shutdown_msg_printed) {
                    shutdown_msg_printed = true;
                }
                return;
            }
        }
    });
    
    if (!batch.empty() && !shutdown_requested_.load()) {
        auto batch_size = batch.size();
        worker_pool_->process_batch(batch);
        state_.total_pairs_generated += batch_size;
        state_.total_pairs_checked += batch_size;
        state_.f_coeffs_state = generator_->get_f_coeffs();
        state_.g_coeffs_state = generator_->get_g_coeffs();
        state_.last_f_index = generator_->get_f_index();
        state_.last_g_index = generator_->get_g_index();
    }

    checkpoint_manager_->force_save();
    
    std::cout << "============================================================" << std::endl;
    std::cout << "FINAL STATISTICS" << std::endl;
    std::cout << "============================================================" << std::endl;
    std::cout << state_.get_summary() << std::endl;
    std::cout << std::endl;
    std::cout << "Note: Results are saved in worker database files." << std::endl;
    std::cout << "Use Python scripts to merge and query results." << std::endl;
}
